### PR TITLE
Add linters to enforce separation between production and test-only crates

### PR DESCRIPTION
**Pull request by sunshowers**
@rexhoffman found an issue with the fuzzing linter #3906 -- it turned out that we weren't considering that crate because it wasn't included in the default build.

Add a bunch of linters to ensure that:
* we have a set of "production crates", which are exactly the set of crates that form publishable artifacts
* everything else is a test-only crate.

### CODE REVIEW

File: documentation/coding_guidelines.md [hunk 1]
```diff
@@ -286,38 +290,70 @@ in tests/benchmarks](https://github.com/rust-lang/cargo/issues/2911), we rely on
...
 2. Annotate your test-only helper `foo` with both the `test` flag (for in-crate callers) and the `"fuzzing"` custom feature (for out-of-crate callers):
-    ```
+    ```rust
     #[cfg(any(test, feature = "fuzzing"))]
     fn foo() { ... }
     ```
 3. (optional) Use `cfg_attr` to make test-only trait derivations conditional:
-    ```
+    ```rust
     #[cfg_attr(any(test, feature = "testing"), derive(FooTrait))]
```
Comment (rexhoffman) [inline documentation/coding_guidelines.md:321]: should we update "testing" to "fuzzing" here?

File: devtools/x-core/src/lib.rs [hunk 2]
```diff
@@ -1,15 +1,28 @@
 // Copyright (c) The Libra Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-use guppy::{graph::PackageGraph, MetadataCommand};
+#[macro_use]
+extern crate rental;
```
Comment (bmwill) [inline devtools/x-core/src/lib.rs:5]: Has the rental crate not been updated to 2018 macros?

File: documentation/coding_guidelines.md (hunk 1 above)
Comment (sunshowers) [reply documentation/coding_guidelines.md:321]: Done!

File: devtools/x-core/src/lib.rs (hunk 2 above)
Comment (sunshowers) [reply devtools/x-core/src/lib.rs:5]: Unfortunately it hasn't been, it still requires this.

**Comment by sunshowers** [comment]
TODO before landing: instead of saying that the test-only list isn't canonical, print out a warning explaining what to do.

File: Cargo.toml [hunk 3]
```diff
@@ -127,14 +127,26 @@ members = [
...
+# never include crates that require fuzzing features or test features. These are the crates we want built with no extra
 # test-only code included.
+#
+# For more, see the "Conditional compilation for tests" section in documentation/coding_guidelines.md.
 default-members = [
+    "config/config-builder",
+    "config/management",
+    "client/libra-dev",
+    "execution/db-bootstrapper",
     "testsuite/cli",
```
Comment (sunshowers) [inline Cargo.toml:140]: TODO before landing: where should this go

**Pull request closed without merging**